### ISSUE TITLE
Unexpected behavior when intercepting AmazonS3Client's `PutObjectRequest`

### ISSUE THREAD

**Issue by ALai57** [biiwide/sandboxica#1]
**CONTEXT:** I am trying to use Sandboxica to intercept PutObject requests to S3 for unit testing.

**PROBLEM STATEMENT:** When I use the `sandboxica/with` method to override the implementation of s3/put-object and I try to call the proxied `put-object` method with a valid `PutObjectRequest`, I expected to be able to intercept the request that I supplied to `put-object`. However, the intercepted request is missing most of the data I tried to send to the `put-object` function (see below).

```
(def s3-req
  (let [some-bytes (.getBytes "<h1>HI</h1>")]
    {:bucket-name    "mybucket"
     :key            "something.txt"
     :input-stream   (java.io.ByteArrayInputStream. some-bytes)
     :metadata       {:content-type   "text/html"
                      :content-length (count some-bytes)
                      :user-metadata  {"somethin" "some-value"}}}))

;; When I override the put-object method on the amazonica s3 client, the
;; intercepted request reflects that the request was not intercepted properly
(sandbox/with (sandbox/just
               (s3/put-object
                ([req] (println "REQUEST" (type req) req))))
  (s3/put-object {:profile "none"} s3-req))
;; => Running this code prints out the following `{:requester-pays? false}`, but is missing all of the supplied fields

```

**WHAT I EXPECTED TO HAPPEN WHEN I RAN THE CODE ABOVE:**
I expected the intercepted request to have the fields I supplied in the request (bucket-name, key, input-stream, metadata... and more). However, the intercepted request doesn't reflect all the fields that were used for inputs

**POSSIBLE PROBLEM:** The missing data in the intercepted request is due to `sandboxica/coerce-method-implementation` missing a preprocessing step before marshalling resources. 

**TRACKING DOWN THE POSSIBLE PROBLEM:** 

```
;;  The `sandboxica/coerce-method-implementation` directly calls `aws/marshall` on the incoming clojure map (`s3-req` in this case)
;;  When `aws/marshall` is called directly on a Clojure map representing PutObjectRequest, it does not marshall resources properly
(aws/marshall (#'aws/create-bean com.amazonaws.services.s3.model.PutObjectRequest
                                 s3-req))
;; => {:requester-pays? false}
```

**EXPLANATION/VERIFICATION:** In the `amazonica.core` library, there is a call to `aws/prepare-args` before resources are marshalled - however this call is not present in the current implementation of `coerce-method-implementation`.
https://github.com/mcohen01/amazonica/blob/master/src/amazonica/core.clj#L867-L887

If `aws/prepare-args` is used before marshalling, the PutObjectRequest seems to marshall properly

```
(let [method
      (#'aws/best-method (:put-object (#'aws/client-methods com.amazonaws.services.s3.AmazonS3Client))
                         [{:profile "none"} s3-req])]
  (bean (first (seq (aws/marshall (#'aws/prepare-args method [s3-req]))))))
;; => {:cloneSource nil,
;;     :storageClass nil,
;;     :key "something.txt",
;;     :customQueryParameters nil,
;;     :file nil,
;;     :readLimit 131073,
;;     :inputStream
;;     #object[java.io.ByteArrayInputStream 0x1f9a0a26 "java.io.ByteArrayInputStream@1f9a0a26"],
;;     :requestCredentials nil,
;;     :objectLockRetainUntilDate nil,
;;     :sdkClientExecutionTimeout nil,
;;     :generalProgressListener
;;     #object[com.amazonaws.event.ProgressListener$NoOpProgressListener 0x775e743b "com.amazonaws.event.ProgressListener$NoOpProgressListener@775e743b"],
;;     :SSECustomerKey nil,
;;     :cloneRoot nil,
;;     :class com.amazonaws.services.s3.model.PutObjectRequest,
;;     :accessControlList nil,
;;     :progressListener nil,
;;     :cannedAcl nil,
;;     :requestClientOptions
;;     #object[com.amazonaws.RequestClientOptions 0x192557b7 "com.amazonaws.RequestClientOptions@192557b7"],
;;     :bucketName "mybucket",
;;     :sdkRequestTimeout nil,
;;     :tagging nil,
;;     :requesterPays false,
;;     :SSEAwsKeyManagementParams nil,
;;     :redirectLocation nil,
;;     :requestMetricCollector nil,
;;     :customRequestHeaders nil,
;;     :metadata
;;     #object[com.amazonaws.services.s3.model.ObjectMetadata 0x58977f3d "com.amazonaws.services.s3.model.ObjectMetadata@58977f3d"],
;;     :requestCredentialsProvider nil,
;;     :objectLockLegalHoldStatus nil,
;;     :objectLockMode nil}

```

**Comment by ALai57** [comment]
Did some more research and have some more clarity on what is going on. 

It seems that the primary issue is that amazonica doesn't have support for marshalling some Classes (such as S3PutRequest). When you call `marshall` on, for example, S3PutRequest, it will dispatch to the `marshall` method for the `Object` data type. This calls `get-fields`, which looks for accessor methods that do not exist for certain S3 objects. Because there are no accessors, marshall won't be able to convert these Java Objects back to Clojure data types, and marshall returns `nil`, or a subset of the data supplied to the AWS client function.

One solution for this is to catch cases where amazonica cannot marshall a particular resource and marshall that resource with custom code. A working attempt compatible with current tests is below - would love input or direction on it.

```
(defn- prop->name
  "Convert java object property of form `isXXX` to be more clojure-y `XXX?`"
  [prop]
  (if (.startsWith prop "is")
    (str (.substring prop 2) "?")
    prop))

(defn aws-coerceable?
  "Does amazonica know how to coerce this data type?"
  [x]
  (get @@#'aws/coercions (class x)))

(defn coerce-to-map
  "Coerce java object to clojure map.
  Used for data types that amazonica doesn't know how to coerce."
  [obj]
  (reduce-kv (fn [m k v]
               (conj m {(#'aws/camel->keyword (prop->name (name k)))
                        (#'aws/marshall v)}))
             {}
             (bean obj)))

(defn- marshall
  [method args]
  (map (fn [x]
         (if (not (aws-coerceable? x))
           (coerce-to-map x)
           x))
       (aws/marshall (#'aws/prepare-args method args))))

(defn coerce-method-implementation
  "Returns a wrapped function that will marshall arguments to Clojure values,
and unmarshall Clojure values to a result type based on the given method signature."
  [f ^Method method]
  (let [rt (.getReturnType method)]
    (fn ([]
         (unmarshall rt (f)))
      ([a]
       (unmarshall rt (apply f (marshall method [a]))))
      ([a b]
       (unmarshall rt (apply f (marshall method [a b]))))
      ([a b & more]
       (unmarshall rt (apply f (marshall method (list* a b more))))))))
```

**Comment by biiwide** [comment]
Looks like Amazonica’s `accessors` function for finding bean property methods is using `getDeclaredMethods` on the class, and all of the interesting methods are on the parent class `AbstractPutObjectRequest`.

**Comment by biiwide** [comment]
@ALai57 try pulling down the `issue/1` branch and see if it helps.

**Comment by ALai57** [comment]
I'll try by this weekend.

Thanks for taking a look! 

There's another (hacky) workaround that I used in the meantime - just extending the `IMarshall` protocol to `PutObjectRequest` manually before running the tests. It's not great, but it does get the job done in a pinch.

**Comment by ALai57** [comment]
Just installed and tried it out - the branch fixed the issue!

**Comment by biiwide** [comment]
Released version `0.3.0` with the fix.
